tests/utils/test_process_util.py

```diff
@@ -32,6 +36,101 @@ def tearDown(self):
         if self.tmp_dir is not None:
             shutil.rmtree(self.tmp_dir)
 
+    def test_wait_for_process_completion_or_timeout_should_terminate_cleanly(self):
+        process = subprocess.Popen(
+            "date",
+            shell=True,
+            cwd=self.tmp_dir,
+            env={},
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE)
+
+        timed_out, ret = wait_for_process_completion_or_timeout(process, 5)
+        self.assertEquals(timed_out, False)
+        self.assertEquals(ret, 0)
+
+    def test_wait_for_process_completion_or_timeout_should_kill_process_on_timeout(self):
+        process = subprocess.Popen(
+            "sleep 4s",
+            shell=True,
+            cwd=self.tmp_dir,
+            env={},
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            preexec_fn=os.setsid)
+
+        with patch('azurelinuxagent.common.utils.processutil.os.killpg') as mock_kill:
+            timed_out, ret = wait_for_process_completion_or_timeout(process, 2)
+
+            self.assertEquals(mock_kill.call_count, 1)
+            self.assertEquals(timed_out, True)
+            self.assertEquals(ret, None)
+
+    def test_wait_for_process_completion_or_timeout_should_return_nonzero_return_code(self):
+        process = subprocess.Popen(
+            "ls folder_does_not_exist",
+            shell=True,
+            cwd=self.tmp_dir,
+            env={},
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE)
+
+        timed_out, ret = wait_for_process_completion_or_timeout(process, 5)
+        self.assertEquals(timed_out, False)
+        self.assertEquals(ret, 2)
+
+    def test_start_subprocess_and_wait_for_completion_should_return_process_output(self):
+        command = "echo 'dummy output'"
+        with tempfile.TemporaryFile(dir=self.tmp_dir, mode="w+b") as stdout:
+            with tempfile.TemporaryFile(dir=self.tmp_dir, mode="w+b") as stderr:
+                process_output = start_subprocess_and_wait_for_completion(command=command,
+                                                                          timeout=5,
+                                                                          shell=True,
+                                                                          cwd=self.tmp_dir,
+                                                                          env={},
+                                                                          stdout=stdout,
+                                                                          stderr=stderr,
+                                                                          preexec_fn=os.setsid,
+                                                                          error_code=42)
+        expected_output = "[stdout]\ndummy output\n\n\n[stderr]\n"
```

Comment:
let's use a test command that produces some stderr

Reply:
Updated test to output to both streams